docs/types/Mutation.md

```diff
@@ -8,7 +8,7 @@ A mutation defines a destructive operation performed on a collection or instance
 
 |   Property   |              Type              |              | Description                                                                                                                                                |
 | :----------: | :----------------------------: | ------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------- |
-|   **type**   |            _string_            | **required** | The type of mutation to perform, must be one of `create`, `update`, or `delete`.                                                                           |
+|   **type**   |            _string_            | **required** | The type of mutation to perform, must be one of `create`, `update`, `jsonPatch`\*, or `delete`.                                                            |
```

Comment:
These types were intentionally kept abstract here (CRUD vs. POST/PUT/PATCH/DELETE/GET) and jsonPatch goes against that intent, as it goes beyond saying _what_ to do by saying _how_ to do it.

The "update" type could be either a PUT or a PATCH under the hood, or, in this case, a "PATCH + json-patch+json".

So conceptually I'm not sure about the `jsonPatch` type. It doesn't belong as a type on the same level as the others.

Reply:
I keep flip-flopping on if I consider json-patch a type of mutation in the same vein as CRUD operations.

Conceptually I want it under the "update" mutation, however the implementation is convincing in its simplicity.

Reply:
So you would prefer `useDataQuery` to remain at its current level of abstraction and hide the format of the data being sent to the server? And I guess if/when we support graphql we also wouldn't want a `type` of graphql so makes sense.

Reply:
@mediremi yeah, that's correct.

Reply:
I can see why you wouldn't want `jsonPatch` as its own type as it doesn't really match with the CRUD types.

But the CRUD types are declarative, which means that the actual implementation is hidden/inaccessible/uncontrollable.
`jsonPatch` on the other hand is an imperative instruction that tells the server what exactly it should do, so my preference would be to not "hide" it behind a declarative type, I think that it would even violate consistency as we'd be mixing declarative CRUD types with an imperative "handle".

When it comes to translating this to the graphql world, then there's nothing that'd stop us from using `jsonPatch` as a type. Graphql only knows query and mutation. What kind of mutations exist depends on the schema. Normally there'd be a `create`/`update`/`delete` mutation for every type (e.g. `createDataElement`, `updateDataElement` and `deleteDataElement`; actually there can be a lot more, like `linkDataElementWithDataSet` for example). There's nothing that prevents us from adding `patchDataElement`/`jsonPatchDataElement`.

Reply:
Right. I think where we see things differently is if the request payload is part of the Data Engine's (our) API or the Server's (their) API.

My perspective is that the payload itself is not part of our API. It's determined by the server in all cases.

So on our side, the developer will declare that they want to do (e.g. update), and pass along a payload. That's declarative as far as our API goes.

If that payload is an ["sequential instruction set"](https://datatracker.ietf.org/doc/html/rfc6902) we pass that along to the server in the proper way.

If that payload is something else (e.g. GQL), we pass that along in the proper way without the user having to think about the specifics of how a request is sent to the server.

In this case, the trade-off is that a developer has to know that we support json-patch instruction sets when using the `update` type. This is consistent with what the developer currently needs to know when s/he works with the API through the Data Engine so I am OK with the trade-off.